### ISSUE TITLE
Fuel node resets connection to indexer

### ISSUE THREAD

**Issue by chlenc** [FuelLabs/fuel-indexer#979]
Hello.

I left the indexer running for a couple of hours on my laptop and when I came back I saw that it wasn't working, here are the latest logs

```
2023-06-05T21:40:10.693653Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: bd9e106f0ab215e8fc8d71fd6b05d7d1a79f780a08640cf2cbbef99059ade9c6 }
2023-06-05T21:40:10.695326Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: bc5c460ddad2560bfd7e9098b838e2fc2a83deb758098b877e58ed82bc22b2b0 }
2023-06-05T21:40:10.696780Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: ba3c4b18159a45fccb776bc784be466958c5302c15a2dd7947b5ad86418e479f }
2023-06-05T21:40:10.698971Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: b8b55b92e899e3b5eb84431aa846ddab5f50e9b85d7ec6da857e0365d44d3986 }
2023-06-05T21:40:10.705034Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: b7f54d0919339925d7fe104bb437addfd13dd06fa2b36cb2c5762f08d9d559ab }
2023-06-05T21:40:10.707353Z  INFO fuel_indexer::ffi: 110: tx = TransactionData { transaction: Script(Script { gas_price: 0, gas_limit: 0, maturity: 0, script: [], script_data: [], inputs: [], outputs: [], witnesses: [], receipts_root: 0000000000000000000000000000000000000000000000000000000000000000, metadata: None }), status: Failure { block: c48237e2a16b4082c894741d82bfdfb2b774188b663d896ce4845be809793b75, time: 1686001208, reason: "Revert(18446744073709486080)", program_state: Some(ProgramState { return_type: Revert, data: b"\0\0\xff\xff\xff\xff\xff\xff" }) }, receipts: [Call { id: 0000000000000000000000000000000000000000000000000000000000000000, to: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, amount: 0, asset_id: 0000000000000000000000000000000000000000000000000000000000000000, gas: 99999685, param1: 1959787426, param2: 14440, pc: 17248, is: 17248 }, LogData { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 0, rb: 12, ptr: 62048, len: 24, digest: a394915e5dfaab8b902cbf199c2688db3d154cbdcb0e32439af016dbc1823e18, data: 4f7264657273207368756f6c64..., pc: 47040, is: 17248 }, Revert { id: 7662a02959e3e2d681589261e95a7a4bc8ac66c6d66999a0fe01bb6c36ada7c6, ra: 18446744073709486080, pc: 47048, is: 17248 }, ScriptResult { result: Revert, gas_used: 4644 }], id: b067509be57c280e1ca5c0149f4a29c01c00bc3effa4b15ffc78228c507ace12 }
2023-06-05T21:40:10.895861Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:12.062910Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:13.229556Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:15.092499Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:16.272329Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:17.495866Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:18.662901Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:19.818391Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:20.972921Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:22.830727Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:23.986512Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:25.141184Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:27.014577Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:28.160126Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:29.323557Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:30.472081Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:31.631490Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:33.429176Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:34.579537Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:35.807554Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:37.614450Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:38.776682Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:40.070312Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:41.228436Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:42.402162Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:43.562428Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:44.719557Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:45.883618Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:47.910900Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:49.067810Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:50.235879Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:51.396923Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:52.544664Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:53.700939Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:54.934098Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:56.085899Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:58.086804Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:40:59.257056Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:00.427419Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:01.646325Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:02.807068Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:03.961312Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:05.116571Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:06.267548Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:08.272058Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:09.426171Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:10.599816Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:11.754234Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:13.632828Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:41:14.793311Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:57:03.513701Z ERROR fuel_indexer::executor: 140: Failed to retrieve blocks: Error making HTTP request: error sending request for url (http://beta-3.fuel.network/graphql): connection error: Connection reset by peer (os error 54)
2023-06-05T21:57:03.534928Z  INFO fuel_indexer::executor: 361: No new blocks to process, sleeping.
2023-06-05T21:57:05.704346Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 1
2023-06-05T21:57:07.311002Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 2
2023-06-05T21:57:08.948136Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 3
2023-06-05T21:57:10.592038Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 4
2023-06-05T21:57:12.242721Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 5
2023-06-05T21:57:13.882789Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 6
2023-06-05T21:57:15.525672Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 7
2023-06-05T21:57:17.154558Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 8
2023-06-05T21:57:18.863931Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
Retrying indexer, attempt: 9
2023-06-05T21:57:20.488126Z ERROR fuel_indexer::executor: 348: Indexer executor failed RuntimeError(RuntimeError { source: Trap(HeapAccessOutOfBounds), wasm_trace: [], native_trace:    0: <unknown>
   1: <unknown>
   2: <unknown>
   3: <unknown>
 }), retrying.
2023-06-05T21:57:21.489565Z ERROR fuel_indexer::executor: 355: Indexer failed after retries, giving up. <('.')>
```

**Comment by ra0x3** [comment]
@chlenc 
- Hmm 🤔 
- I see that your indexer was processing blocks, then, for whatever reason, the fuel node stopped sending blocks
- Indexers are currently configured to just pass nothing to the indexer handler if there is an error fetching blocks
  - [That logic is here](https://github.com/FuelLabs/fuel-indexer/blob/master/packages/fuel-indexer/src/executor.rs#L143)
- I'm wondering if fuel-core is sending weird output? 

CC @deekerno

**Comment by deekerno** [comment]
Yeah, this is definitely on the `fuel-core` side.

"Connection reset by peer" is the result of the other party in a TCP/IP handshake sending you a RST packet; think of it as the client not wanting to send you any more packets for some reason and they're being "polite" by letting you know instead of ghosting you. In any case, I don't know what circumstances lead to such a thing on the deployment side; perhaps it's a rate-limiting measure?

@rfuelsh, @tirkesi: Any thoughts here?

**Comment by ra0x3** [comment]
cc @Voxelot as well. 
- @Voxelot wondering
  1. What causes the blocks to stop being produced (seemingly intermittently?)
  2. Could fuel-client be sending something we don't expect, thus causing their indexer modules to break ☝🏽 
    - Because if the indexer fails to talk to fuel-client, we just send empty results, so that wouldn't break things

**Comment by Voxelot** [comment]
I think the blocks are still getting produced, but maybe fuel-core / nginx terminated the connection for some reason. We'll need to investigate this on the devops / client team side, but in the meantime it might be good to implement a reconnect mechanism on the indexer if the connection gets dropped by the server.

**Comment by ra0x3** [comment]
https://github.com/FuelLabs/fuel-indexer/issues/984

**Comment by chlenc** [comment]
GM ☀️
are any updates here?

**Comment by chlenc** [comment]
> 

Let's have a call?

**Comment by chlenc** [comment]
GM frens
Are any updates here? 
Is there any sense in trying to deploy again or is it definitely not working yet?

**Comment by ra0x3** [comment]
@chlenc Yes, update is:
- Unfortunately, this `Trap(HeapAccessOutOfBounds)` error is an error primarily on the _indexer_ level, not the Fuel Indexer service level
  - Meaning, there is most likely some code in your indexer causing this error to occur
- With regard to the  `Failed to retrieve blocks: Error making HTTP request` error
  - This is an error yes, but it's an error that fails gracefully
    - If we can't make the request to the fuel client, we just pass an empty array of blocks (as you mentioned)
      - This shouldn't cause the `Trap(HeapAccessOutOfBounds)` error for an empty array
        -  But even then, we've added some logic to never call `handle_events` on empty blocks ([here](https://github.com/FuelLabs/fuel-indexer/blob/develop/packages/fuel-indexer/src/executor.rs#L784))

- TLDR:
  - I would try your indexer again on the newest version 0.19.2 (which is being released now) and respond here with any updates on your end

**Comment by ra0x3** [comment]
Gonna close this now, but feel free to reopen if the issue persists (or open a new discussion)
